File: app/src/main/java/d/d/meshenger/AddressUtils.java
/*
 * Generate possible IP addresses  to reach a contact.
*/
package d.d.meshenger;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


public class AddressUtils {
    private static final String TAG = "AddressUtils";

    public static InetSocketAddress[] getAllSocketAddresses(List<String> initial_addresses, InetAddress last_working_address, int port) {
        Set<InetSocketAddress> address_set = new HashSet<>();

        if (last_working_address != null) {
            address_set.add(new InetSocketAddress(last_working_address, port));
        }

        for (String address : initial_addresses) {
            try {
                if (Utils.isMAC(address)) {
                    // use own addresses as template
                    address_set.addAll(getAddressPermutations(address, port));
                    // from neighbor table
                    address_set.addAll(getAddressesFromNeighborTable(address, port));
                } else {
                    // parse address
                    address_set.add(InetSocketAddress.createUnresolved(address, port));
                }
            } catch (Exception e) {
                Log.e(TAG, "Invalid address: " + address);
                e.printStackTrace();
            }
        }

        // sort addresses
        InetSocketAddress[] address_array = address_set.toArray(new InetSocketAddress[0]);
        Arrays.sort(address_array, (InetSocketAddress lhs, InetSocketAddress rhs) -> {
            InetAddress rhs_addr = lhs.getAddress();
            InetAddress lhs_addr = rhs.getAddress();

            // prefer last working address
            if (last_working_address != null) {
                if (last_working_address.getAddress().equals(lhs_addr)) {
                    return -1;
                }
                if (last_working_address.getAddress().equals(rhs_addr)) {
                    return 1;
                }
            }

            if (lhs.isUnresolved() && !rhs.isUnresolved()) {
                return -1;
            }
            if (!lhs.isUnresolved() && rhs.isUnresolved()) {
                return 1;
            }

            int lhs_proto = getAddressProtocolValue(lhs_addr);
            int rhs_proto = getAddressProtocolValue(rhs_addr);

            if (lhs_proto < rhs_proto) {
                return -1;
            }
            if (lhs_proto > rhs_proto) {
                return 1;
            }

            int lhs_scope = getAddressScopeValue(lhs_addr);
            int rhs_scope = getAddressScopeValue(rhs_addr);

            if (lhs_scope < rhs_scope) {
                return -1;
            } else if (lhs_scope > rhs_scope) {
                return 1;
            } else {
                return 0;
            }
        });

        for (InetSocketAddress address : address_array) {
            Log.d(TAG, "got address: " + address);
        }

        return address_array;
    }

    public static List<AddressEntry> getOwnAddresses() {
        ArrayList<AddressEntry> addressList = new ArrayList<>();
        try {
            List<NetworkInterface> all = Collections.list(NetworkInterface.getNetworkInterfaces());
            for (NetworkInterface nif : all) {
                byte[] mac = nif.getHardwareAddress();

                if (nif.isLoopback()) {
                    continue;
                }

                if (Utils.isMAC(mac)) {
                    addressList.add(new AddressEntry(Utils.bytesToMacAddress(mac), nif.getName(), Utils.isMulticastMAC(mac)));
                }

                for (InterfaceAddress ia : nif.getInterfaceAddresses()) {
                    InetAddress addr = ia.getAddress();
                    if (addr.isLoopbackAddress()) {
                        continue;
                    }

                    addressList.add(new AddressEntry(addr.getHostAddress(), nif.getName(), addr.isMulticastAddress()));
                }
            }
        } catch (Exception ex) {
            // ignore
            Log.d(TAG, "error: " + ex.toString());
        }

        return addressList;
    }

    private static int getAddressProtocolValue(InetAddress addr) {
        if (addr instanceof Inet6Address) {
            return 1;
        } else {
            return 2;
        }
    }

    private static int getAddressScopeValue(InetAddress addr) {
        if (addr.isLoopbackAddress()) {
            return 0;
        }

        if (addr.isLinkLocalAddress()) {
            return 1;
        }

        if (addr.isSiteLocalAddress()) {
            return 2;
        }

        if (addr.isMulticastAddress()) {
            if (addr.isMCGlobal()) {
                return 3;
            }
            if (addr.isMCLinkLocal()) {
                return 4;
            }
            if (addr.isMCNodeLocal()) {
                return 5;
            }
            if (addr.isMCOrgLocal()) {
                return 6;
            }
            if (addr.isMCSiteLocal()) {
                return 7;
            }
        }

        return 8;
    }


    // list all IP/MAC addresses of running network interfaces
    // for debugging only
    public static void printOwnAddresses() {
        for (AddressEntry ae : getOwnAddresses()) {
            Log.d(TAG, "Address: " + ae.address + " (" + ae.device + (ae.multicast ? ", multicast" : "") + ")");
        }
    }

    // Check if the given MAC address is in the IPv6 address
    static byte[] getEUI64MAC(Inet6Address addr6) {
        byte[] bytes = addr6.getAddress();

        // check for EUI-64 address
        if (bytes[11] != ((byte) 0xFF) || bytes[12] != ((byte) 0xFE)) {
            return null;
        }

        byte[] mac = new byte[6];
        mac[0] = (byte) (bytes[8] ^ 2);
        mac[1] = bytes[9];
        mac[2] = bytes[10];
        mac[3] = bytes[13];
        mac[4] = bytes[14];
        mac[5] = bytes[15];

        return mac;
    }

    private static List<InetSocketAddress> getAddressesFromNeighborTable(String lookup_mac, int port) {
        List<InetSocketAddress> addrs = new ArrayList<>();
        try {
        	// get IPv4 and IPv6 entries
            Process pc = Runtime.getRuntime().exec("ip n l");
            BufferedReader rd = new BufferedReader(
                new InputStreamReader(pc.getInputStream(), "UTF-8")
            );

            String line = "";
            while ((line = rd.readLine()) != null) {
                String[] tokens = line.split("\\s+");
                // IPv4
                if (tokens.length == 6) {
                    String addr = tokens[0];
                    String device = tokens[2];
                    String mac = tokens[4];
                    String state = tokens[5];
                    if (lookup_mac.equalsIgnoreCase(mac) && Utils.isIP(addr) && !state.equalsIgnoreCase("failed")) {
                        if (addr.startsWith("fe80:") || addr.startsWith("169.254.")) {
                            addrs.add(new InetSocketAddress(addr + "%" + device, port));
                        } else {
                            addrs.add(new InetSocketAddress(addr, port));
                        }
                    }
                }

                // IPv6
                if (tokens.length == 7) {
                    String addr = tokens[0];
                    String device = tokens[2];
                    String mac = tokens[4];
                    String state = tokens[6];
                    if (mac.equalsIgnoreCase(lookup_mac) && Utils.isIP(addr) && !state.equalsIgnoreCase("failed")) {
                        if (addr.startsWith("fe80:") || addr.startsWith("169.254.")) {
                            addrs.add(new InetSocketAddress(addr + "%" + device, port));
                        } else {
                            addrs.add(new InetSocketAddress(addr, port));
                        }
                    }
                }
            }
        } catch (IOException e) {
            Log.d(TAG, e.toString());
        }

        return addrs;
    }

    /*
    * Replace the MAC address of an EUi64 scheme IPv6 address with another MAC address.
    * E.g.: ("fe80::aaaa:aaff:faa:aaa", "bb:bb:bb:bb:bb:bb") => "fe80::9bbb:bbff:febb:bbbb"
    */
    private static Inet6Address createEUI64Address(Inet6Address addr6, byte[] mac) {
        // addr6 is expected to be a EUI64 address
        try {
            byte[] bytes = addr6.getAddress();

            bytes[8] = (byte) (mac[0] ^ 2);
            bytes[9] = mac[1];
            bytes[10] = mac[2];

            // already set, but doesn't harm
            bytes[11] = (byte) 0xFF;
            bytes[12] = (byte) 0xFE;

            bytes[13] = mac[3];
            bytes[14] = mac[4];
            bytes[15] = mac[5];

            return Inet6Address.getByAddress(null, bytes, addr6.getScopeId());
        } catch (UnknownHostException e) {
            return null;
        }
    }

    /*
    * Iterate all own addresses of the device. Check if they conform to the EUI64 scheme.
    * If yes, replace the MAC address in it with the supplied one and return that address.
    * Also set the given port for those generated addresses.
    */
    private static List<InetSocketAddress> getAddressPermutations(String contact_mac, int port) {
        byte[] contact_mac_bytes = Utils.macAddressToBytes(contact_mac);
        ArrayList<InetSocketAddress> addrs = new ArrayList<InetSocketAddress>();
        try {
            List<NetworkInterface> all = Collections.list(NetworkInterface.getNetworkInterfaces());
            for (NetworkInterface nif : all) {
                if (nif.isLoopback() || nif.getName().equals("dummy0")) {
                    continue;
                }

                for (InterfaceAddress ia : nif.getInterfaceAddresses()) {
                    InetAddress addr = ia.getAddress();
                    if (addr.isLoopbackAddress()) {
                        continue;
                    }

                    if (addr instanceof Inet6Address) {
                        Inet6Address addr6 = (Inet6Address) addr;
                        byte[] extracted_mac = getEUI64MAC(addr6);
                        if (extracted_mac != null && Arrays.equals(extracted_mac, nif.getHardwareAddress())) {
                            // We found the interface MAC address in the IPv6 address assigned to that interface in the EUI-64 scheme.
                            // Now assume that the contact has an address with the same scheme.
                            InetAddress new_addr = createEUI64Address(addr6, contact_mac_bytes);
                            if (new_addr != null) {
                                addrs.add(new InetSocketAddress(new_addr, port));
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return addrs;
    }
}

File: app/src/main/kotlin/d/d/meshenger/AddressUtils.kt
/*
* Copyright (C) 2025 Meshenger Contributors
* SPDX-License-Identifier: GPL-3.0-or-later
*/

package d.d.meshenger

import android.net.InetAddresses
import android.os.Build
import android.util.Patterns
import d.d.meshenger.MainService.Companion.DEFAULT_PORT
import d.d.meshenger.MainService.Companion.MAX_PORT
import d.d.meshenger.MainService.Companion.MIN_PORT
import java.net.*
import java.util.*
import java.util.regex.Pattern
import kotlin.experimental.xor

internal object AddressUtils
{
    fun closeSocket(socket: Socket?) {
        try {
            socket?.close()
        } catch (_: Exception) {
            // ignore
        }
    }

    fun isValidPort(portString: String?): Boolean {
        return portString != null && parsePort(portString) != -1
    }

    fun parsePort(portString: String, defaultPort: Int = -1): Int {
        var port = -1
        try {
            port = portString.toInt()
        } catch (_: Exception) {
            // ignore parse error
        }

        return if (port in MIN_PORT..MAX_PORT) {
            port
        } else {
            defaultPort
        }
    }

    // List any interfaces we do not want to display
    // because we know that is cannot be used.
    fun ignoreDeviceByName(device: String): Boolean {
        return device.startsWith("dummy")
    }

    fun formatMAC(mac: ByteArray): String {
        return String.format("%02X:%02X:%02X:%02X:%02X:%02X",
            mac[0], mac[1], mac[2], mac[3], mac[4], mac[5])
    }

    /*
     * Strip device name from address:
     * "192.168.1.1%wlan1" => "192.168.1.1"
     * "fe80::1%wlan0" => "fe80::1"
     * "[fe80::1%wlan0]:1234" => "[fe80::]:1234"
     */
    fun stripInterface(address:String): String {
        val beg = address.indexOf('%')

        if (beg != -1) {
            val end = address.indexOfAny(charArrayOf(']', ':'), startIndex = beg)
            return if (end == -1) {
                address.substring(0, beg)
            } else {
                address.substring(0, beg) + address.substring(end)
            }
        }

        return address;
    }

    /*
     * Strip first entry, e.g.:
     * "/1.2.3.4" => "1.2.3.4"
     * "google.com/1.2.3.4" => "1.2.3.4"
     * "google.com" => "google.com"
     */
    fun stripHost(address: String): String {
        val pos = address.indexOf('/')
        return address.substring(pos + 1)
    }

    fun stripPort(address: String): String {
        val pos = address.indexOf("]:")
        if (pos != -1) {
            address.substring(1, pos + 1)
        }

        if (address.count { it == ':' } == 1) {
            val pos = address.lastIndexOf(":")
            return if (pos != -1) {
                address.substring(0, pos)
            } else {
                address
            }
        }

        return address
    }

    // remove interface, host and port from address
    fun stripAddress(address: String): String {
        return stripHost(stripPort(stripInterface(address)))
    }

    // coarse address type distinction
    enum class AddressType {
        GLOBAL_IP, LOCAL_IP, MULTICAST_IP, DOMAIN
    }

    fun parseInetAddress(address: String): InetAddress? {
        if (isIPAddress(address)) {
            val inetAddress = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                InetAddresses.parseNumericAddress(address)
            } else {
                InetAddress.getByName(address)
            }
            return inetAddress
        }

        return null
    }

    fun getAddressType(address: String): AddressType {
        val ipAddress = parseInetSocketAddress(address)
        if (ipAddress != null) {
            if (ipAddress.address.isMulticastAddress) {
                return AddressType.MULTICAST_IP
            } else if ((ipAddress.address.address[1].toInt() and 15) == 0x0E) {
                // global IP address
                return AddressType.GLOBAL_IP
            } else {
                return AddressType.LOCAL_IP
            }
        }

        return AddressType.DOMAIN
    }

    private fun isValidMACBytes(mac: ByteArray?): Boolean {
        // we ignore the first byte (dummy mac addresses have the "local" bit set - resulting in 0x02)
        return (mac != null
                && mac.size == 6
                && mac[1].toInt() != 0x0
                && mac[2].toInt() != 0x0
                && mac[3].toInt() != 0x0
                && mac[4].toInt() != 0x0
                && mac[5].toInt() != 0x0)
    }

    private val DOMAIN_PATTERN = Pattern.compile("^[a-zA-Z0-9\\-._]{1,255}$")
    private val MAC_PATTERN = Pattern.compile("^[0-9a-fA-F]{2}(:[0-9a-fA-F]{2}){5}$")

    private val DEVICE_PATTERN = Pattern.compile("^[a-zA-Z0-9]{1,8}$")
    private val IPV6_STD_PATTERN = Pattern.compile("^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$")
    private val IPV6_HEX_COMPRESSED_PATTERN = Pattern.compile("^((?:[0-9a-fA-F]{1,4}(?::[0-9a-fA-F]{1,4})*)?)::((?:[0-9a-fA-F]{1,4}(?::[0-9a-fA-F]{1,4})*)?)$")

    fun isIPAddress(address: String): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            return InetAddresses.isNumericAddress(address)
        } else {
            @Suppress("DEPRECATION")
            return IPV6_STD_PATTERN.matcher(address).matches()
                 || IPV6_HEX_COMPRESSED_PATTERN.matcher(address).matches()
                 || Patterns.IP_ADDRESS.matcher(address).matches()
        }
    }

    fun isMACAddress(address: String): Boolean {
        return MAC_PATTERN.matcher(address).matches()
    }

    fun isDomain(address: String): Boolean {
        val addrWithoutPort = address.substringBefore(':')
        return DOMAIN_PATTERN.matcher(addrWithoutPort).matches()
            && !addrWithoutPort.contains("..")
    }

    // replacement of isIPAddress
    fun isInetSocketAddress(address: String?): Boolean {
        return (null != parseInetSocketAddress(address))
    }

    /* Parses an address string:
     *
     * "[<ip-address>]:<port>"
     * "[<ip-address>]"
     * "<ip-address>:<port>"
     * "<ip-address>"
     * "<domain>:<port>"
     * "<domain>"
    */
    fun parseInetSocketAddress(address: String?, defaultPort: Int = DEFAULT_PORT): InetSocketAddress? {
        if (address.isNullOrEmpty()) {
            return null
        } else if (address.startsWith("[")) {
            val end = address.lastIndexOf("]:")
            if (end > 0) {
                // "[<ip-address>]:<port>"
                val addressPart = address.substring(1, end)
                val portPart = address.substring(end + 2)
                val port = parsePort(portPart)
                if (port != null && isIPAddress(addressPart)) {
                    return InetSocketAddress(addressPart, port)
                }
            } else {
                val end = address.lastIndexOf("]")
                // [<ip-address>]
                if (end > 0) {
                    val addressPart = address.substring(1, end)
                    if (isIPAddress(addressPart)) {
                        return InetSocketAddress(addressPart, defaultPort)
                    }
                }
            }
        } else {
            if (address.count { it == ':' } == 1) {
                // "<domain>:<port>"
                // "<ip-address>:<port>"
                val end = address.indexOf(":")
                val addressPart = address.substring(0, end)
                val portPart = address.substring(end + 1)
                val port = parsePort(portPart)
                if (port != -1) {
                    return if (isIPAddress(addressPart)) {
                        InetSocketAddress(addressPart, port)
                    } else {
                        InetSocketAddress.createUnresolved(addressPart, port)
                    }
                }
            } else if (isIPAddress(address)) {
                // "<ip-address>"
                return InetSocketAddress(address, defaultPort)
            } else if (isDomain(address)) {
                // "<domain>"
                return InetSocketAddress.createUnresolved(address, defaultPort)
            }
        }
        return null
    }

    fun inetSocketAddressToString(socketAddress: InetSocketAddress?): String? {
        if (socketAddress == null) {
            return null
        }
        val address = socketAddress.address
        val port = socketAddress.port
        if (port !in MIN_PORT..MAX_PORT) {
            return null
        } else if (address == null) {
            val host = socketAddress.hostString.trimStart('/')
            if (isIPAddress(host) || isDomain(host)) {
                return "$host:$port"
            } else {
                return null
            }
        } else if (address is Inet6Address) {
            val str = stripHost(address.toString())
            return "[$str]:$port"
        } else {
            val str = stripHost(address.toString())
            return "$str:$port"
        }
    }

    fun collectAddresses(): List<AddressEntry> {
        val addressList = ArrayList<AddressEntry>()
        try {
            for (nif in Collections.list(NetworkInterface.getNetworkInterfaces())) {
                if (nif.isLoopback) {
                    continue
                }

                if (ignoreDeviceByName(nif.name)) {
                    continue
                }

                for (ia in nif.interfaceAddresses) {
                    if (ia.address.isLoopbackAddress) {
                        continue
                    }

                    val hostAddress = ia.address.hostAddress
                    if (hostAddress != null) {
                        addressList.add(AddressEntry(stripInterface(hostAddress), nif.name))
                    }
                }
            }
        } catch (ex: Exception) {
            // ignore
            Log.d(this, "collectAddresses() error=$ex")
        }

        return addressList.distinct()
    }

    // list all IP/MAC addresses of running network interfaces - for debugging only
    fun printOwnAddresses() {
        for (ae in collectAddresses()) {
            val type = getAddressType(ae.address)
            Log.d(this, "Address: ${ae.address} (${ae.device} $type)")
        }
    }

    /*
    * Replace the MAC address of an EUi64 scheme IPv6 address with another MAC address.
    * E.g.: ("fe80::aaaa:aaff:faa:aaa", "bb:bb:bb:bb:bb:bb") => "fe80::9bbb:bbff:febb:bbbb"
    */
    fun createEUI64Address(address: Inet6Address, mac: ByteArray): Inet6Address {
        // address is expected to be an EUI64 address
        val bytes = address.address
        bytes[8] = (mac[0] xor 2)
        bytes[9] = mac[1]
        bytes[10] = mac[2]
        // ff:fe may or mac not be already set
        bytes[11] = 0xFF.toByte()
        bytes[12] = 0xFE.toByte()
        bytes[13] = mac[3]
        bytes[14] = mac[4]
        bytes[15] = mac[5]
        return Inet6Address.getByAddress(null, bytes, address.scopeId)
    }

    /*
    * E.g. "11:22:33:44:55:66" => "fe80::1322:33ff:fe44:5566"
    */
    fun getLinkLocalFromMAC(macAddress: String): String? {
        val bytes = macAddressToBytes(macAddress)
        val inetAddress = parseInetAddress("fe80::")
        if (bytes != null && inetAddress is Inet6Address) {
            val inetAddressEUI64 = createEUI64Address(inetAddress, bytes)
            return inetAddressEUI64.toString().substring(1)
        }
        return null
    }

    /*
    * E.g. "11:22:33:44:55:66" => [0x11, 0x22, 0x33, 0x44, 0x55, 0x66]
    */
    fun macAddressToBytes(macAddress: String): ByteArray? {
        if (isMACAddress(macAddress)) {
            val bytes = macAddress
                .split(":")
                .map { it.toInt(16).toByte() }
                .toByteArray()
            if (isValidMACBytes(bytes)) {
                return bytes
            }
        }
        return null
    }
}
